src/packages/auth/src/utils/get-rules-for-roles.ts

```diff
@@ -0,0 +1,48 @@
+import {
+	AccessType,
+	BASE_ROLE_EVERYONE,
+	MultiFactorAuthentication,
+	MultiFactorAuthenticationOperationType,
+	MultiFactorAuthenticationRule,
+} from '../types';
+
+export const getRulesForRoles = (
+	mfa: MultiFactorAuthentication, // The MFA rules for this entity
+	roles: string[], // The roles assigned by the current user
+	operation: AccessType // The operation performed by the user
+): MultiFactorAuthenticationRule[] => {
+	const rules = [] as MultiFactorAuthenticationRule[];
+
+	// Let's loop through the rules for this entity and get a list of those that apply to the current users roles
+	for (const role of [...roles, BASE_ROLE_EVERYONE]) {
+		const operations = mfa[role];
+
+		// Check if we have a matching rule for this operation
+		if (operations?.[operation]) {
+			const rule = operations[operation];
+			if (rule) rules.push(...rule);
+		}
+
+		// Check if we have a rule that matches a write operation
+		const writeOperations = [
+			MultiFactorAuthenticationOperationType.CREATE,
+			MultiFactorAuthenticationOperationType.UPDATE,
+			MultiFactorAuthenticationOperationType.DELETE,
+		];
+		if (
+			writeOperations.includes(operation as any) &&
```

Comment:
It's marginal, but we could make this more efficient with:
```typescript
const writeOperations = new Set([
			MultiFactorAuthenticationOperationType.CREATE,
			MultiFactorAuthenticationOperationType.UPDATE,
			MultiFactorAuthenticationOperationType.DELETE,
		]);
		if (
			writeOperations.has(operation as any) &&
```

And even better if we do the `new Set` at the top of the file instead of reinitialising it on each loop.